File: examples/chatapp/node-be/src/resolvers/query.ts
import * as db from "../db.ts";
import type { UserRow } from "../db.ts";
import { mapDb } from "../errors.ts";
import type { QueryResolvers } from "../generated/graphql.ts";
import { parseId, requireAuth, requireMember } from "./helpers.ts";

export const Query: QueryResolvers = {
  me: async (_r, _a, ctx) => {
    if (!ctx.currentUser) return null;
    try {
      return await ctx.loaders.userById.load(ctx.currentUser.id);
    } catch (e) {
      throw mapDb(e);
    }
  },

  chats: async (_r, _a, ctx) => {
    const user = requireAuth(ctx);
    try {
      return await db.chatsForUser(ctx.app.pool, user.id);
    } catch (e) {
      throw mapDb(e);
    }
  },

  chat: async (_r, { id }, ctx) => {
    const user = requireAuth(ctx);
    const chatId = parseId(id);
    await requireMember(ctx.app, chatId, user.id);
    try {
      return await db.chatById(ctx.app.pool, chatId);
    } catch (e) {
      throw mapDb(e);
    }
  },

  messages: async (_r, { chatId, before, limit }, ctx) => {
    const user = requireAuth(ctx);
    const id = parseId(chatId);
    await requireMember(ctx.app, id, user.id);
    const lim = Math.min(Math.max(limit, 1), 200);
    try {
      return await db.listMessages(ctx.app.pool, id, before ?? null, lim);
    } catch (e) {
      throw mapDb(e);
    }
  },

  presence: async (_r, { userIds }, ctx) => {
    requireAuth(ctx);
    const ids = userIds.map(parseId);
    let rows: UserRow[];
    try {
      rows = await db.usersByIds(ctx.app.pool, ids);
    } catch (e) {
      throw mapDb(e);
    }
    // Preserve request order, drop unknown ids.
    const byId = new Map(rows.map((u) => [u.id, u]));
    return ids.flatMap((id) => {
      const u = byId.get(id);
      return u ? [u] : [];
    });
  },

  reactionsEnabled: async (_r, _a, ctx) => {
    if (!ctx.currentUser) return false;
    return ctx.app.reactionsEnabled(ctx.currentUser.id);
  },

  opsStats: async (_r, _a, ctx) => {
    requireAuth(ctx);
    const [onlineCount, dlqCount] = await Promise.all([
      ctx.app.onlineCount(),
      ctx.app.dlqCount(),
    ]);
    return { onlineCount, dlqCount };
  },
};

File: examples/chatapp/rust-be/src/gql/query.rs
use async_graphql::{Context, ID, Object, Result};
use chrono::{DateTime, Utc};

use crate::context::CurrentUser;
use crate::db;
use crate::gql::helpers::{app, load_user, map_db, me, parse_id, require_member};
use crate::gql::types::{GqlChat, GqlMessage, GqlUser, OpsStats};

pub struct Query;

#[Object]
impl Query {
    async fn me(&self, ctx: &Context<'_>) -> Result<Option<GqlUser>> {
        let Some(user) = ctx.data_opt::<CurrentUser>() else {
            return Ok(None);
        };
        Ok(load_user(ctx, user.id).await.ok().map(GqlUser))
    }

    async fn chats(&self, ctx: &Context<'_>) -> Result<Vec<GqlChat>> {
        let c = app(ctx)?;
        let user = me(ctx)?;
        Ok(db::chats_for_user(&c.pool, user.id)
            .await
            .map_err(map_db)?
            .into_iter()
            .map(GqlChat)
            .collect())
    }

    async fn chat(&self, ctx: &Context<'_>, id: ID) -> Result<Option<GqlChat>> {
        let c = app(ctx)?;
        let user = me(ctx)?;
        let chat_id = parse_id(&id)?;
        require_member(c, chat_id, user.id).await?;
        Ok(db::chat(&c.pool, chat_id)
            .await
            .map_err(map_db)?
            .map(GqlChat))
    }

    async fn messages(
        &self,
        ctx: &Context<'_>,
        chat_id: ID,
        before: Option<DateTime<Utc>>,
        #[graphql(default = 50)] limit: i32,
    ) -> Result<Vec<GqlMessage>> {
        let c = app(ctx)?;
        let user = me(ctx)?;
        let cid = parse_id(&chat_id)?;
        require_member(c, cid, user.id).await?;
        let limit = limit.clamp(1, 200) as i64;
        Ok(db::list_messages(&c.pool, cid, before, limit)
            .await
            .map_err(map_db)?
            .into_iter()
            .map(GqlMessage)
            .collect())
    }

    async fn presence(&self, ctx: &Context<'_>, user_ids: Vec<ID>) -> Result<Vec<GqlUser>> {
        let c = app(ctx)?;
        me(ctx)?;
        let mut ids = Vec::with_capacity(user_ids.len());
        for id in &user_ids {
            ids.push(parse_id(id)?);
        }
        Ok(db::users_by_ids(&c.pool, &ids)
            .await
            .map_err(map_db)?
            .into_iter()
            .map(GqlUser)
            .collect())
    }

    async fn reactions_enabled(&self, ctx: &Context<'_>) -> Result<bool> {
        let Some(user) = ctx.data_opt::<CurrentUser>() else {
            return Ok(false);
        };
        Ok(app(ctx)?.reactions_enabled(user.id).await)
    }

    async fn ops_stats(&self, ctx: &Context<'_>) -> Result<OpsStats> {
        let c = app(ctx)?;
        me(ctx)?;
        Ok(OpsStats {
            online_count: c.online_count().await as i32,
            dlq_count: c.dlq_count().await as i32,
        })
    }
}
